File: src/save_solution.py
import numpy as np
import csv

def save_solution(filename, P):
    """
    Saves a permutation matrix P to a CSV solution file.
    """
    # find() in scipy.sparse is equivalent to MATLAB's find
    rows, cols = P.nonzero()
    
    # Convert back to 1-based indexing for the output file
    male_nodes = rows + 1
    female_nodes = cols + 1
    
    with open(filename, 'w', newline='') as f:
        writer = csv.writer(f)
        writer.writerow(['Male Node ID', 'Female Node ID'])
        for i, j in zip(male_nodes, female_nodes):
            writer.writerow([f'm{i}', f'f{j}'])

File: src_m/save_solution.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

function save_solution(filename,P)

[i,j] = find(P);
fid = fopen(filename,'w');
fprintf(fid,'Male Node ID,Female Node ID\n'); 
fprintf(fid,'m%d,f%d\n',[i j]');              
fclose(fid);

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
